### ISSUE TITLE
Set multiple Environment

### ISSUE THREAD

**Issue by nickp4488** [petercinibulk/envied#109]
Hello,
I have set up my variable like API Key and base url in .env file to secure. It works perfectly fine.

I want to differentiate between DEV,TEST,PROD environment, I am not sure should I make 3 different .env file or how can I set up based of 3 different flavours.

I have seen issues #91  & #63 but unable to understand clearly.

Please add something into ReadMe file or guide how can we achieve this?

Thank you,

**Comment by techouse** [comment]
@nickp4488 check the [example](https://github.com/petercinibulk/envied/tree/main/examples/envied_example)

**Comment by nickp4488** [comment]
I already seen example but I have multiple environment like debug, stage, test, prod. Is there any facility to provide ENV file path from command line?

**Comment by a1tem** [comment]
Hi @techouse I also checked the example and I can't figure out how to use it, I also checked other threads could you please update readme.md file with clear example how people can use it.

Thanks!

**Comment by a1tem** [comment]
Okay, I figured this out:

1) Create 3 env files
```
.env.dev
.env.staging
.env.prod
```

Add corresponding values to those files:
`API_URL=https://bla-bla-bla.net/api`

Next, create inside lib/env folders:
**app_env_fields.dart**
```
abstract interface class AppEnvFields {
  abstract final String apiUrl;
}
```

**app_env.dart**
```
import 'dev_env.dart';
import 'staging_env.dart';
import 'prod_env.dart';

import 'app_env_fields.dart';

abstract interface class AppEnv implements AppEnvFields {
  static const String? appFlavor =
      String.fromEnvironment('FLUTTER_APP_FLAVOR') != ''
          ? String.fromEnvironment('FLUTTER_APP_FLAVOR')
          : null;

  factory AppEnv() => _instance;

  static final AppEnv _instance = appFlavor == 'dev'
      ? DevEnv()
      : appFlavor == 'staging'
          ? StagingEnv()
          : appFlavor == 'prod'
              ? ProdEnv()
              : throw Exception('Invalid app flavor');
}
```

**dev_env.dart**
```
import 'app_env.dart';
import 'app_env_fields.dart';
import 'package:envied/envied.dart';

part 'dev_env.g.dart';

@Envied(name: 'Env', path: '.env.dev')
final class DevEnv implements AppEnv, AppEnvFields {
  DevEnv();

  @override
  @EnviedField(varName: 'API_URL')
  final String apiUrl = _Env.apiUrl;
}

```
**staging_env.dart**
```
import 'app_env.dart';
import 'app_env_fields.dart';
import 'package:envied/envied.dart';

part 'staging_env.g.dart';

@Envied(name: 'Env', path: '.env.staging')
final class StagingEnv implements AppEnv, AppEnvFields {
  StagingEnv();

  @override
  @EnviedField(varName: 'API_URL')
  final String apiUrl = _Env.apiUrl;
}

```

**prod_env.dart**
```
import 'app_env.dart';
import 'app_env_fields.dart';
import 'package:envied/envied.dart';

part 'prod_env.g.dart';

@Envied(name: 'Env', path: '.env.prod')
final class ProdEnv implements AppEnv, AppEnvFields {
  ProdEnv();

  @override
  @EnviedField(varName: 'API_URL')
  final String apiUrl = _Env.apiUrl;
}
```

After that in any file in the application, you can use it like this:
```
AppEnv().apiUrl
```
for example in dio provider:
```
import 'core/utils/interceptors/error_interceptor.dart';
import 'env/app_env.dart';
import 'package:dio/dio.dart';
import 'package:riverpod_annotation/riverpod_annotation.dart';
import 'core/utils/interceptors/logger_interceptor.dart';

part 'dio_provider.g.dart';

@riverpod
Dio dio(DioRef ref) {
  print('API URL: ${AppEnv().apiUrl}');
  BaseOptions options = BaseOptions(
    baseUrl: AppEnv().apiUrl,
  );

  final dio = Dio(options);
  dio.interceptors.add(LoggerInterceptor());
  dio.interceptors.add(ErrorInterceptor());
  return dio;
}

```

**Comment by techouse** [comment]
@a1tem correct, just like the example :)

**Comment by tenhobi** [comment]
@a1tem @techouse Hi, why do you suggest this approach? With this you have to have all .env.dev, .env.staging, .env.prod files even in local development, no? It also means, that you generate .g file for all of these files. And if you do some selection like this

```dart
  static final AppEnv _instance = appFlavor == 'dev'
      ? DevEnv()
      : appFlavor == 'staging'
          ? StagingEnv()
          : appFlavor == 'prod'
              ? ProdEnv()
              : throw Exception('Invalid app flavor');
```

it will probably be also bundled with the app, no? Because tree shaking cannot shake it?

I wanted to resolve it with using appFlavor from Flutter, but even through it is a const, it cannot be used in annotations lol. 

```dart
@Envied(name: 'Env', path: '.env.${appFlavor}')
final class Env {}
```

So for now I just have to use just one `.env` file and if I want to switch to i.e. prod build locally, I have to replace it with different values or link to prod file from it, which is I guess ok.

---

TL;DR with suggested setup, how do you make sure other environments (dev, staging, prod, ..., qa, whatever) are not build with the mainly prod app?
